Split libATen.so into libATen_cpu.so and libATen_cuda.so

ezyang opened the pull request:
Stacked on top of https://github.com/pytorch/pytorch/pull/7197

Previously, ATen could be built with either CPU-only support, or
CPU/CUDA support, but only via a compile-time flag, requiring
two separate builds.  This means that if you have a program which
indirectly uses a CPU-only build of ATen, and a CPU/CUDA-build of
ATen, you're gonna have a bad time.  And you might want a CPU-only
build of ATen, because it is 15M (versus the 300M of a CUDA build).

This commit splits libATen.so into two libraries, CPU/CUDA, so
that it's not necessary to do a full rebuild to get CPU-only
support; instead, if you link against libATen_cpu.so only, you
are CPU-only; if you additionally link/dlopen libATen_cuda.so,
this enables CUDA support.  This brings ATen's dynamic library
structure more similar to Caffe2's.

The general principle for how this works is that we introduce
a *hooks* interface, which introduces a dynamic dispatch indirection
between a call site and implementation site of CUDA functionality,
mediated by a static initialization registry.  This means that we can continue
to, for example, lazily initialize CUDA from Context (a core, CPU class) without
having a direct dependency on the CUDA bits.  Instead, we look up
in the registry if, e.g., CUDA hooks have been loaded (this loading
process happens at static initialization time), and if they
have been we dynamic dispatch to this class.  We similarly use
the hooks interface to handle Variable registration.

We introduce a new invariant: if the backend of a type has not
been initialized (e.g., it's library has not been dlopened; for
CUDA, this also includes CUDA initialization), then the Type
pointers in the context registry are NULL.  If you access the
registry directly you must maintain this invariant.

We also preserve libATen.so, which is an empty stub library that
depends on libATen_cpu.so and libATen_cuda.so, depending on
the setting of USE_CUDA.

There are a few potholes along the way.  I document them here:

- Previously, PyTorch maintained a separate registry for variable
  types, because no provision for them was made in the Context's
  type_registry.  Now that we have the hooks mechanism, we can easily
  have PyTorch register variables in the main registry.  The code
  has been refactored accordingly.

- There is a subtle ordering issue between Variable and CUDA.
  We permit libATen_cuda.so and PyTorch to be loaded in either
  order (in practice, CUDA is always loaded "after" PyTorch, because
  it is lazily initialized.)  This means that, when CUDA types are
  loaded, we must subsequently also initialize their Variable equivalents.
  Appropriate hooks were added to VariableHooks to make this possible;
  similarly, getVariableHooks() is not referentially transparent, and
  will change behavior after Variables are loaded.  (This is different
  to CUDAHooks, which is "burned in" after you try to initialize CUDA.)

- The cmake is adjusted to separate dependencies into either CPU
  or CUDA dependencies.  The generator scripts are adjusted to either
  generate a file as a CUDA (cuda_file_manager) or CPU file (file_manager).

- I changed all native functions which were CUDA-only (the cudnn functions)
  to have dispatches for CUDA only (making it permissible to not specify
  all dispatch options.)  This uncovered a bug in how we were handling
  native functions which dispatch on a Type argument; I introduced a new
  self_ty keyword to handle this case.  I'm not 100% happy about it
  but it fixed my problem.

  This also exposed the fact that set_history incompletely handles
  heterogenous return tuples combining Tensor and TensorList.  I
  swapped this codegen to use flatten() (at the possible cost of
  a slight perf regression, since we're allocating another vector now
  in this code path).

- thc_state is no longer a public member of Context; use getTHCState() instead

- This PR comes with Registry from Caffe2, for handling static initialization.

ezyang committed:
Split libATen.so into libATen_cpu.so and libATen_cuda.so

Previously, ATen could be built with either CPU-only support, or
CPU/CUDA support, but only via a compile-time flag, requiring
two separate builds.  This means that if you have a program which
indirectly uses a CPU-only build of ATen, and a CPU/CUDA-build of
ATen, you're gonna have a bad time.  And you might want a CPU-only
build of ATen, because it is 15M (versus the 300M of a CUDA build).

This commit splits libATen.so into two libraries, CPU/CUDA, so
that it's not necessary to do a full rebuild to get CPU-only
support; instead, if you link against libATen_cpu.so only, you
are CPU-only; if you additionally link/dlopen libATen_cuda.so,
this enables CUDA support.  This brings ATen's dynamic library
structure more similar to Caffe2's.  libATen.so is no more
(this is BC BREAKING)

The general principle for how this works is that we introduce
a *hooks* interface, which introduces a dynamic dispatch indirection
between a call site and implementation site of CUDA functionality,
mediated by a static initialization registry.  This means that we can continue
to, for example, lazily initialize CUDA from Context (a core, CPU class) without
having a direct dependency on the CUDA bits.  Instead, we look up
in the registry if, e.g., CUDA hooks have been loaded (this loading
process happens at static initialization time), and if they
have been we dynamic dispatch to this class.  We similarly use
the hooks interface to handle Variable registration.

We introduce a new invariant: if the backend of a type has not
been initialized (e.g., it's library has not been dlopened; for
CUDA, this also includes CUDA initialization), then the Type
pointers in the context registry are NULL.  If you access the
registry directly you must maintain this invariant.

There are a few potholes along the way.  I document them here:

- Previously, PyTorch maintained a separate registry for variable
  types, because no provision for them was made in the Context's
  type_registry.  Now that we have the hooks mechanism, we can easily
  have PyTorch register variables in the main registry.  The code
  has been refactored accordingly.

- There is a subtle ordering issue between Variable and CUDA.
  We permit libATen_cuda.so and PyTorch to be loaded in either
  order (in practice, CUDA is always loaded "after" PyTorch, because
  it is lazily initialized.)  This means that, when CUDA types are
  loaded, we must subsequently also initialize their Variable equivalents.
  Appropriate hooks were added to VariableHooks to make this possible;
  similarly, getVariableHooks() is not referentially transparent, and
  will change behavior after Variables are loaded.  (This is different
  to CUDAHooks, which is "burned in" after you try to initialize CUDA.)

- The cmake is adjusted to separate dependencies into either CPU
  or CUDA dependencies.  The generator scripts are adjusted to either
  generate a file as a CUDA (cuda_file_manager) or CPU file (file_manager).

- I changed all native functions which were CUDA-only (the cudnn functions)
  to have dispatches for CUDA only (making it permissible to not specify
  all dispatch options.)  This uncovered a bug in how we were handling
  native functions which dispatch on a Type argument; I introduced a new
  self_ty keyword to handle this case.  I'm not 100% happy about it
  but it fixed my problem.

  This also exposed the fact that set_history incompletely handles
  heterogenous return tuples combining Tensor and TensorList.  I
  swapped this codegen to use flatten() (at the possible cost of
  a slight perf regression, since we're allocating another vector now
  in this code path).

- thc_state is no longer a public member of Context; use getTHCState() instead

- This PR comes with Registry from Caffe2, for handling static initialization.
  I needed to make a bunch of fixes to Registry to make it more portable

  - No more ##__VA_ARGS__ token pasting; instead, it is mandatory to pass at
    least one argument to the var-args. CUDAHooks and VariableHooks pass a nullary
    struct CUDAHooksArgs/VariableHooksArgs to solve the problem. We must get rid of
    token pasting because it does not work with MSVC.

  - It seems MSVC is not willing to generate code for constructors of template
    classes at use sites which cross DLL boundaries. So we explicitly instantiate
    the class to get around the problem. This involved tweaks to the boilerplate
    generating macros, and also required us to shuffle around namespaces a bit,
    because you can't specialize a template unless you are in the same namespace as
    the template.
  - Insertion of AT_API to appropriate places where the registry must be exported

- We have a general problem which is that on recent Ubuntu distributions,
  --as-needed is enabled for shared libraries, which is (cc @apaszke who was
  worrying about this in #7160 see also #7160 (comment)). For now, I've hacked
  this up in the PR to pass -Wl,--no-as-needed to all of the spots necessary to
  make CI work, but a more sustainable solution is to attempt to dlopen
  libATen_cuda.so when CUDA functionality is requested.

    - The JIT tests somehow manage to try to touch CUDA without loading libATen_cuda.so. So
      we pass -Wl,--no-as-needed when linking libATen_cuda.so to _C.so

- There is a very subtle linking issue with lapack, which is solved by making sure libATen_cuda.so links against LAPACK. There's a comment in aten/src/ATen/CMakeLists.txt about htis as well as a follow up bug at #7353

- autogradpp used AT_CUDA_ENABLED directly. We've expunged these uses and added
  a few more things to CUDAHooks (getNumGPUs)

- Added manualSeedAll to Generator so that we can invoke it polymorphically (it
  only does something different for CUDAGenerator)

- There's a new cuda/CUDAConfig.h header for CUDA-only ifdef macros (AT_CUDNN_ENABLED, most prominently)

- CUDAHooks/VariableHooks structs live in at namespace because Registry's
  namespace support is not good enough to handle it otherwise (see Registry
  changes above)

- There's some modest moving around of native functions in ReduceOps and
  UnaryOps to get the CUDA-only function implementations into separate files, so
  they are only compiled into libATen_cuda.so. sspaddmm needed a separate CUDA
  function due to object linkage boundaries.

- Some direct uses of native functions in CUDA code has to go away, since these
  functions are not exported, so you have to go through the dispatcher
  (at::native::empty_like to at::empty_like)

- Code in THC/THCS/THCUNN now properly use THC_API macro instead of TH_API
  (which matters now that TH and THC are not in the same library)

- Added code debt in torch/_thnn/utils.py and other THNN parsing code to handle
  both TH_API and THC_API

- TensorUtils.h is now properly exported with AT_API

- Dead uses of TH_EXPORTS and co expunged; we now use ATen_cpu_exports and
  ATen_cuda_exports (new, in ATenCUDAGeneral.h) consistently

- Fix some incorrect type annotations on _cudnn_rnn_backward, where we didn't
  declare a type as possibly undefined when we should have. We didn't catch this
  previously because optional annotations are not tested on "pass-through" native
  ATen ops (which don't have dispatch). Upstream issue at #7316

- There's a new cmake macro aten_compile_options for applying all of our
  per-target compile time options. We use this on the cpu and cuda libraries.

- test/test_cpp_extensions.py can be run directly by invoking in Python,
  assuming you've setup your PYTHONPATH setup correctly

- type_from_string does some new funny business to only query for all valid CUDA
  types (which causes CUDA initialization) when we see "torch.cuda." in the
  requested string

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented on torch/csrc/tensor/python_tensor.cpp:
```diff
@@ -176,8 +194,10 @@ static THPObjectPtr get_storage_obj(const Type& type) {
 static void set_type(PyTensorType& type_obj, Backend backend, ScalarType scalarType) {
-  auto baseType = globalContext().type_registry[static_cast<int>(backend)][static_cast<int>(scalarType)].get();
-  type_obj.aten_type = baseType ? torch::autograd::VariableType::getType(*baseType) : nullptr;
+  // This field is lazily initialized from backend and scalar_type
```
So, strictly speaking, the problem is if you call set_type before you initialize CUDA (and not just whether or not you loaded ATen_cuda.so). And in this case, calling set_type before initialization basically always happens, because at program initialization time, CUDA is not yet initialized.

ezyang commented on aten/src/ATen/cuda/detail/CUDAHooks.h:
```diff
@@ -0,0 +1,26 @@
+  void registerCUDATypes(Context*) const override;
+};
+
+// Sigh, the registry doesn't support namespaces :(
```
Yes that would work, although ideally a straight `REGISTER_CUDA_HOOKS` would just work.

ezyang replied:
Oh, we don't actually want to do this, because then the key name from the registry is "cuda::detail::CUDAHooks" eek!

ezyang commented on aten/src/ATen/detail/CUDAHooksInterface.cpp:
```diff
@@ -0,0 +1,18 @@
+
+namespace at { namespace detail {
+
+const CUDAHooksInterface& getCUDAHooks() {
```
Yes it does. I also fixed the construction in VariableHooks

ezyang commented:
@ebetica The functions I added should be semantics preserving for what the Torch++ API was previously. I agree that Context's mismash of functions are a little inconsistent. Can we just "do what PyTorch does" to solve the API discrepancy?

ebetica commented:
It seems to be what autogradpp does: https://pytorch.org/docs/master/_modules/torch/cuda.html#is_available

ezyang commented:
Here is English updates on all of the bugfixes on this PR:

- We have a general problem which is that on recent Ubuntu distributions, `--as-needed` is enabled for shared libraries, which is (cc @apaszke who was worrying about this in https://github.com/pytorch/pytorch/pull/7160 see also https://github.com/pytorch/pytorch/pull/7160#issuecomment-387216823). For now, I've hacked this up in the PR to pass `-Wl,--no-as-needed` to all of the spots necessary to make CI work, but a more sustainable solution is to attempt to `dlopen` `libATen_cuda.so` when CUDA functionality is requested.
  - (UNDER TESTING) The JIT tests somehow manage to try to touch CUDA without loading `libATen_cuda.so`. I'm testing a fix which is passing `-Wl,--no-as-needed` when linking `libATen_cuda.so` to `_C.so`
- I needed to make a bunch of fixes to `Registry` to make it more portable
  - No more `##__VA_ARGS__` token pasting; instead, it is mandatory to pass at least one argument to the var-args.  CUDAHooks and VariableHooks pass a nullary struct CUDAHooksArgs/VariableHooksArgs to solve the problem. We must get rid of token pasting because it does not work with MSVC.
  - It seems MSVC is not willing to generate code for constructors of template classes at use sites which cross DLL boundaries. So we explicitly instantiate the class to get around the problem. This involved tweaks to the boilerplate generating macros, and also required us to shuffle around namespaces a bit, because you can't specialize a template unless you are in the same namespace as the template.
- There is a very subtle linking issue with lapack, which is solved by making sure `libATen_cuda.so` links against LAPACK. There's a comment in `aten/src/ATen/CMakeLists.txt` about htis as well as a follow up bug at https://github.com/pytorch/pytorch/issues/7353
- There is no more `libATen.so` compatibility library. You must use `libATen_cpu.so` or `libATen_cuda.so`. A lot of adjustments to CPP extensions, libtorch, etc. were made to make this go through correctly. This will need fbcode fixes as well.
- autogradpp used `AT_CUDA_ENABLED` directly. We've expunged these uses and added a few more things to `CUDAHooks` (`getNumGPUs)`
- Added `manualSeedAll` to `Generator` so that we can invoke it polymorphically (it only does something different for CUDAGenerator)
- There's a new `cuda/CUDAConfig.h` header for CUDA-only ifdef macros (`AT_CUDNN_ENABLED`, most prominently)
- CUDAHooks/VariableHooks structs live in `at` namespace because Registry's namespace support is not good enough to handle it otherwise (see Registry changes above)
- There's some modest moving around of native functions in ReduceOps and UnaryOps to get the CUDA-only function implementations into separate files, so they are only compiled into `libATen_cuda.so`. `sspaddmm` needed a separate CUDA function due to object linkage boundaries.
- Some direct uses of native functions in CUDA code has to go away, since these functions are not exported, so you have to go through the dispatcher (`at::native::empty_like` to `at::empty_like`)
- Code in THC/THCS/THCUNN now properly use `THC_API` macro instead of `TH_API` (which matters now that TH and THC are not in the same library)
  - Some code debt in `torch/_thnn/utils.py` and other THNN parsing code to handle both `TH_API` and `THC_API`
- `TensorUtils.h` is now properly exported with `AT_API`
- Dead uses of `TH_EXPORTS` and co expunged

ezyang commented:
@pytorchbot retest this please

ezyang commented:
@pytorchbot retest this please

apaszke commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -262,31 +284,49 @@ set(TBB_BUILD_TESTS OFF CACHE BOOL " " FORCE)
-endif(NOT ${CMAKE_VERSION} VERSION_LESS "3.1")
+  set_property(TARGET ATen_cpu PROPERTY CXX_STANDARD 11)
+  if(CUDA_FOUND)
+    set_property(TARGET ATen_cuda PROPERTY CXX_STANDARD 11)
```
Shouldn't this be guarded by `NOT NO_CUDA` as well?

ezyang replied:
I always forget which one I'm supposed to test for. I think the CMakeLists.txt is also inconsistent about it. Will fix.

apaszke commented on aten/src/ATen/common_with_cwrap.py:
```diff
@@ -190,6 +190,13 @@ def parse_header(path):
+                fn_name = fn_name[1:-2]
+            else:
+                fn_name = fn_name[:-1]
+            generic_functions.append(Function(fn_name))
```
Why do we need this now?

ezyang replied:
I dunno if we need it or not, but I was fixing places where we hardcoded `TH_API` (because I flipped a bunch of these to be `THC_API`)

apaszke commented on aten/src/ATen/detail/CUDAHooksInterface.h:
```diff
@@ -0,0 +1,114 @@
+//  1. Does the *implementation* of this function require linking against
+//     CUDA libraries?
+//
+//  2. Is this function *called* from non-CUDA ATen code?
```
I think those two points are not enough to justify having all the functions you put there. Why would CPU-only code care how many CUDA devices there are? Why would it ever want to get a current CUDA stream?

Also, having min eps for batch norm in here is very weird too. We should just have a single entry point for CUDA dispatch in the CPU library, and handle choosing a CUDA impl internally in `ATen_cuda`. Ofc it doesn't have to happen in this PR, and I guess C10 might resolve the issue anyway.

ezyang replied:
Right, so I think eventually we want way less functions in this interface. When I put things in here, I didn't bother "changing" any of ATen's API; if there was a "CPU" part of ATen which needed some CUDA information, I tossed it in. I think that many of the examples you provide can be eliminated.  I absolutely agree with your point about batchnorm min-epsilon.

I do want to push on the matter of "should ATen have a getNumGPUs function" a little more. I think there's a pretty strong case to be made for putting making an "ATen CUDA interface", which you can only link against if you actually link against `libATen_cuda.so`. However, I don't think it is a 100% slam dunk. Let's think about an average Python script which supports both CPU and CUDA execution. The same Python script works no matter if you run it with PyTorch compiled with CPU+CUDA, or PyTorch with CPU only. The reason, of course, is that you don't actually exercise the CUDA functionality unless you *actually call it*, so as long as you don't pass `--cuda` to the trainer script it will happily use the CPU functionality. Now let's consider the corresponding situation in C++ land. If you link against a `libATen_cuda.so` only symbol, you MUST have an install of PyTorch which has `libATen_cuda.so` available (in all its 300M glory) (which in turn must have `libcudart.so`), UNLESS we distribute some sort of `stub/libATen_cuda.so` which defines all of the public symbols of libATen_cuda.so but with trivial implementations of them. On the other hand, if the symbols were defined in `libATen.so` and dynamically dispatched to `libATen_cuda.so`, everything would work perfectly fine. Now, to be clear, I am not sure what the correct decision here is, but I don't think it is trivial.

ezyang replied:
(Blah, actually, batchnorm min-epsilon is not so easy to expunge: the problem is that the current dispatch logic doesn't know how dispatch first to CUDA, and *then* to the autograd-enabled sub-implementation. If I give anything `dispatch: CUDA` then I immediately lose pass through and have to implement the backwards manually. Don't want to do that.)

apaszke commented on aten/src/ATen/detail/VariableHooksInterface.h:
```diff
@@ -0,0 +1,42 @@
+
+// The VariableHooksInterface is an interface for autograd functionality
+// which currently doesn't live in libATen.so AND needs to be called from
+// ATen.  In this case, it is only the type registry for Variable types,
```
Why do we use VariableType anywhere in ATen?

ezyang replied:
Yeah, it's used to add Variable variants of types whenever we register new types to the type registry.  The specific case we have to worry about looks like this:

1. libATen.so is loaded, and it initializes the type registry with CPU types.  However, CUDA types are not initialized (they are only initialized when we lazily initialize CUDA)
2. libtorch.so is loaded, and it initializes the variable type registry with all the types in the type registry, which in this case is only the CPU types
3. During execution, CUDA initialization occurs.  CUDA initializes the type registry with CUDA types, and now we need to also, at the same time we're initializing the CUDA types, also initialize the CUDA variable types.

Also, because C10 is going to have native knowledge about Variables, I sucked the variable type registry into ATen.

apaszke commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -89,11 +89,18 @@ def TypedDict(name, attrs, total=True):  # type: ignore
 ${return_type} ${Type}::${api_name}(${type_method_formals}) const {
-    ${return_call} at::native::${native_type_method_dispatch}(${actuals});
+    const auto& self_ty = *this;
+    (void)self_ty;
```
Is this `self_ty` part used anywhere? I can't find relevant code

ezyang replied:
This is a heinous hack to make `_cudnn_init_dropout_state` work:

```
- func: _cudnn_init_dropout_state(Type self_ty, double dropout, bool train, int64_t dropout_seed) -> Tensor
  variants: function
  dispatch:
    CUDA: _cudnn_init_dropout_state
```

It is the sole native function which dispatches based on a Type argument rather than a Tensor argument.  The codegen seems to know how to do the dispatch on `self_ty` (yay), but then it gets stuck when making the call to `_cudnn_init_dropout_state` because it needs to refer to the `Type` that we dynamically dispatched on, but it's not in scope.  Thus the `self_ty` binding.  I'm all ears for better ways to implement it.

apaszke commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -44,29 +43,23 @@ Tensor batch_norm(
-
-#if AT_CUDNN_ENABLED()
-  if (use_cudnn && eps >= CUDNN_BN_MIN_EPSILON) {
+               && detail::getCUDAHooks().compiledWithCuDNN()
```
Would be nice if we avoided calling `getCUDAHooks` too many times. After all every call is an atomic load or some kind of atomic test and set for `call_once`

ezyang replied:
On a TSO memory model (e.g., X86) an atomic load is just a plain old MOV...

apaszke commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -44,29 +43,23 @@ Tensor batch_norm(
+               && cudnn_enabled && detail::getCUDAHooks().versionCuDNN() >= 5110L);
+
+  if (use_cudnn && eps >= detail::getCUDAHooks().batchnormMinEpsilonCuDNN()) {
     return std::get<0>(at::cudnn_batch_norm(
```
Actually how does that work? Will dynamically loading `ATen_cuda` magically change this symbol? Or do we depend on lazy binding to make sure that this path will simply never be taken if `ATen_cuda` is not loaded, and won't cause missing symbol errors because of that?

ezyang replied:
Not quite.

First, batchnormMinEpsilonCuDNN is a virtual function call defined in libATen_cpu.so, so we never have to worry about symbol resolution; the symbol is always defined, it's simply a matter of what object/vtable we are actually making the virtual function call on.  So how do we get our hands on the object?  We're using the Caffe2 registry, but what this effectively is, is a global variable which gets modified by the static initializer that runs when we dlopen('libATen_cuda.so')

OK, but as you say, can we just 'magically change the symbol'?  Here things can get a bit complicated.  This patch makes a simplifying assumption: once you make a call to ANY function that would require CUDA initialization, and it turns out that 'libATen_cuda.so' has NOT been loaded at this point, then we (1) give an error and (2) *permanently disable CUDA for the process.*

This is not intrinsic to the design, but it's easy to get the locking correct in this scheme.  One elaboration on this scheme, not for this PR, but Soumith/Yangqing/I agreed we should do, is have lazy CUDA initialization attempt to dlopen('libATen_cuda.so') *itself*, so that if you really do have CUDA it won't really be possible to get this wrong.

apaszke commented on aten/src/TH/THGeneral.h.in:
```diff
@@ -29,7 +29,7 @@
 
 #ifdef _WIN32
-# ifdef TH_EXPORTS
+# ifdef ATen_cpu_EXPORTS
```
lol

apaszke commented on setup.py:
```diff
@@ -566,27 +566,33 @@ def run(self):
-ATEN_LIB = os.path.join(lib_path, 'libATen.so')
+ATEN_LIBS = [os.path.join(lib_path, 'libATen_cpu.so')]
+if WITH_CUDA:
+    ATEN_LIBS.extend(['-Wl,--no-as-needed', os.path.join(lib_path, 'libATen_cuda.so'), '-Wl,--as-needed'])
```
This won't work on mac. Not all linkers support `--no-as-needed`

ezyang replied:
But it's OK because this line is Linux only anyway (see below for the dylib path)

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -465,10 +465,8 @@ def get_return_value():
-        else:
-            outs = CodeTemplate("{ ${outs} }").substitute(outs=output_names)
+        # TODO: flatten allocates a std::vector, which could be expensive
+        outs = CodeTemplate("flatten( ${outs} )").substitute(outs=output_names)
```
Why do we need this now?

ezyang replied:
Previously, `_cudnn_rnn_backward`, which returns a heterogenous tuple of `(Tensor, Tensor, TensorList)`, did not have a `dispatch` and did not have an entry in `derivatives.yaml`, so IF you attempt to double-backwards through cuDNN RNN it would have attempted to take the derivative of all the internal calls in `_cudnn_rnn_backward` (including, e.g., `data_ptr`) which obviously isn't going to work.  Since I changed `_cudnn_rnn_backward` to be dispatched and only available on CUDA, this flipped the codepath to now treat the insides of `_cudnn_rnn_backward` as nondifferentiable and set history. And *that* triggered the fact that history setting on the return types didn't work with heterogenous tuples.  Thus the change here.

apaszke commented on tools/autograd/templates/VariableType.cpp:
```diff
@@ -251,6 +270,18 @@ static std::vector<SavedVariable> make_saved_variable_list(TensorList tensors) {
+  for (auto& t : tl) {
+    variables.emplace_back(make_variable(std::move(t), /*requires_grad=*/false));
+  }
+  return variables;
```
Use `fmap`. You missed `reserve`, and that would do it for you

ezyang committed:
Last mile libtorch fixes

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
pedantic fix

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
Yay, everything passed!!

apaszke reviewed (commented):
Haven't reviewed the CMake and the file manager parts. The rest looks ok. It's not the prettiest patch, but I hope it will go away soon, and things will be much cleaner in c10.

ezyang commented on aten/src/ATen/gen.py:
```diff
@@ -124,7 +122,16 @@ def check_all_files_written(self):
+                      [static_cast<int>(Backend::${backend})]
+                      [static_cast<int>(ScalarType::${scalar_type})]
+                      .reset(new ${type_name}(context));
+detail::getVariableHooks().registerVariableTypeFor(context, Backend::${backend}, ScalarType::${scalar_type});
```
Nope, either order works. If ATen is loaded first, then libtorch reads out the types which are registered in ATen and initializes based on it. If libtorch is loaded first, ATen has a hook call in its initializers that gives libtorch an opportunity to add variable types.

ezyang commented on aten/src/ATen/Context.h:
```diff
@@ -45,20 +57,44 @@ class AT_API Context {
-      doInitCUDA();
+      thc_state = detail::getCUDAHooks().initCUDA();
+      generator_registry[static_cast<int>(Backend::CUDA)] =
+        detail::getCUDAHooks().initCUDAGenerator(this);
```
Yes, it's equivalent to the old code (contents of `doInitCUDA` inlined into this function)

ezyang commented:
I agree with the other CI comments but I am going to merge to avoid a CI turnaround and I'll post a follow up PR.
